XCP-D update (#259)
DCAN-Labs/cdni-brain: docs/pipelines.md

# Pipelines

This page has the recommended flags and resources for the major pipelines we use for data processing as well as example commands. To determine the exact resources needed to run each pipeline, see [the optimizing resources page](optimizing.md). Below is a table outlining the recommended pipeline combinations. 

<table>
  <tr>
   <td>
<strong>Infants</strong>
   </td>

>    <td>infant-fMRIprep -> XCP-D

Reviewer: ```suggestion
   <td>infant-fMRIprep (NiBabies) -> XCP-D
```

   </td>
  </tr>
  <tr>
   <td><strong>Children (≥4 years) and Adults</strong>
   </td>
   <td>fMRIPrep -> XCP-D
   </td>
  </tr>
  <tr>
   <td><strong>Legacy Data</strong>
   </td>

>    <td>ABCD-BIDS -> XCP-D

Reviewer: ```suggestion
   <td>ABCD-BIDS (abcd-hcp-pipeline) -> XCP-D
```

   </td>
  </tr>
</table>

Note: the cutoff age for using an adult pipeline depends on the scope of the project. 

<div class="admonition attention">
    <p class="first admonition-title">Attention</p>
    <p class="last">
        ABCD-BIDS should only be used as a preprocessing pipeline when wanting to compare your dataset with other legacy data that was processed that way (i.e. ABCD and HCP). Any newly collected datasets should be processed with fMRIprep and XCP-D.
    </p>
</div>
  

Recommended sbatch parameters per 1 subject 1 session process:

<table>
<tr>
<td>
<strong>Pipeline</strong>
</td>
<td><strong>Recommended SBATCH Parameters </strong>
</td>
</tr>
<tr>
<td>fMRIprep
</td>
<td>24 hours, 32 cpus, with 240gb of total memory. 
<p>
</td>
</tr>
<tr>
<td>infant-fMRIprep
</td>
<td>24 hours, 32 cpus, with 240gb of total memory. 
<p>
<em>May need increased resources for data without precomputed derivatives for segmentation (if JLF segmentation is enabled with `--segmentation-atlases-dir` )</em>
</td>
</tr>
<tr>
<td>abcd-bids-pipeline
</td>

> <td> 48 hours, 8 tasks on one node (1 cpu per task), 20gb of memory and 100gb of temporary storage.

Reviewer: ```suggestion
<td> 48 hours, 8 cpus, 80gb of memory, 100gb of temporary storage.
```

</td>
</tr>
<tr>
<td>XCP-D
</td>
<td>3 hours, 8 cpus, and 96gb of memory.
</td>
</tr>
<tr>
<td>nhp-10.5T-abcd-bids-pipeline
</td>

> <td>24 hours for one run of a subject, add another 4 hours per run. <br /> At least 8 cpus, max amount is twice the number of functional runs. <br /> 10 times the number of CPUs is the total memory needed.

Reviewer: ```suggestion
<td>48 hours minimum, up to 96 hours if session length is long (more than an hour of functional data). <br /> At least 8 cpus, max amount is twice the number of functional runs. <br /> 10gb memory per CPU recommended.
```

<p>
</td>
</tr>
<tr>
<td>dcan-infant
</td>

> <td>72 hours, 24 tasks on one node, 60gb of memory and 100gb of temporary storage.

Reviewer: ```suggestion
<td>72 hours, 8 cpus, 80gb of memory and 100gb of temporary storage. Consider increasing memory to 20gb per cpu or more if functional run lengths are long (more than 10 minutes).
```

    
    * `--nprocs 32` : maximum number of threads across all processes **(if used, this should match the number of cpus allotted for your sbatch job)**
    
    * `--omp-nthreads 3` : maximum number of threads per-process
    
    * `--resource-monitor` : enable Nipype’s resource monitoring to keep track of memory and CPU usage
    
    * `-vv` : level 2 verbose log output, useful for troubleshooting
    
    * `-w /work` : used to specify a working directory within the same path as the sbatch, named_ /work_.

2. If an error is encountered, check for a solution on the [troubleshooting page](troubleshooting.md#nibabies-and-fmriprep) and consider creating a [Github issue](https://github.com/DCAN-Labs/cdni-brain/issues) or adding it yourself if your error is not there.

3. Example command:

```
module load singularity

singularity='which singularity'

${singularity} run --cleanenv \
-B ${fmriprep_in}/input:/data:ro \
-B ${fmriprep_out}/bcp_fmriprep_outputs:/out \
-B /home/faird/shared/code/external/utilities/freesurfer_license/license.txt:/opt/freesurfer/license.txt \
-B /tmp:/work \

> /home/faird/shared/code/external/pipelines/fmriprep/fmriprep_23.0.1.sif /data/out participant \

Reviewer: ```suggestion
/home/faird/shared/code/external/pipelines/fmriprep/fmriprep_24.1.1.sif /data/out participant \
```

-w /work
```

## 2. Infant-fMRIprep (aka NiBabies)

Read: [NiBabies ReadTheDocs site](https://nibabies.readthedocs.io/en/latest/)

NiBabies is a robust pre-processing MRI and fMRI workflow that is also a part of the NiPreps family. NiBabies is designed and optimized for human infants between 0-2 years old. For more information on NiBabies code, see [here](https://github.com/nipreps/nibabies). For in-depth usage information, see the [Read the Docs here](https://nibabies.readthedocs.io/en/latest/usage.html).

1. Preferred flags:

    * `--participant-label \` : a space delimited list of participant identifiers or a single identifier (the sub- prefix can be removed)
    

>     * `--age-months \`: used to specify the age in months of the participant that is being processed 

Reviewer: ```suggestion
    * `--age-months \`: used to specify the age in months of the participant that is being processed. (NOTE: this option is planned to be made deprecated; the recommended method to specify age is with a BIDS sessions or participants file: https://nibabies.readthedocs.io/en/latest/usage.html#participant-ages)
```

    
    * `-vv \` : level 2 verbose log output, useful for troubleshooting
    
    * `--nprocs 32 \` : maximum number of threads across all processes
    
    * `--omp-nthreads 3 \` : maximum number of threads per-process
    
    * `-w /work` : used to specify a working directory within the container’s filesystem, named _/work_.

2. If an error is encountered, check for a solution on the [troubleshooting page](troubleshooting.md#nibabies-and-fmriprep) and consider creating a [Github issue](https://github.com/DCAN-Labs/cdni-brain/issues) or adding it yourself if your error is not there.

3. Example command:

```
module load singularity

singularity='which singularity'

${singularity} run --cleanenv \
-B ${nibabies_in}/input/:/data:ro \
-B ${nibabies_out}/nibabies_outputs/:/out \
-B /home/faird/shared/code/external/utilities/freesurfer_license/license.txt:/opt/freesurfer/license.txt \
-B ${tmp_dir}/work_dir/:/work \
-B ${ext_derivs}/bibsnet_outputs/:/derivatives \

> /home/faird/shared/code/external/pipelines/nibabies/nibabies_24.0.0.sif /data /out participant \

Reviewer: ```suggestion
/home/faird/shared/code/external/pipelines/nibabies/nibabies_24.1.0.sif /data /out participant \
```

--participant-label {subject_ID} \ 
--age-months {age} \

> -s {age_mo} \

Reviewer: ```suggestion
--session-id {session_ID} \
```

--derivatives /derivatives \
--cifti-output 91k \
--omp-nthreads 3 \
-w /work
```

> ## 3. ABCD-HCP-BIDS

Reviewer: ```suggestion
## 3. ABCD-BIDS (abcd-hcp-pipeline)
```

    </p>
</div>

XCP-D versions 0.8.0 and above have a new required `mode` flag that will set several defaults. The preferred mode to use for processing within the center is the `abcd` mode, which sets the default flags explained below. Any of these inputs can be overridden if necessary, but give careful thought as to why they would need to be different for your dataset.

1. Default flags:
    
    * `--fd-thresh/-f 0.3` : framewise displacement threshold for censoring.
    
    * `--file-format cifti` : postprocess cifti instead of NIfTI; this is set default for dcan and hcp input
    
    * `--warp-surfaces-native2std` : resample surfaces to fsLR space, 32k density, and apply the transform from native T1w to MNI152NLin6Asym space output by fMRIPrep / NiBabies
    
    * `--abcc-qc` to output an Executive Summary and HDF5 file containing motion levels at different thresholds

    * `--linc-qc` to output QC metrics from the LINC pipeline

    * `-m`/`--combine-runs` : add concatenated outputs for functional derivatives

    * `--despike` : despike the NIfTI/cifti before postprocessing

>     * `--input-type fmriprep` : automatically assumes inputs are from fMRIprep

Reviewer: ```suggestion
    * `--input-type fmriprep` : automatically assumes inputs are from fMRIprep (other options are `dcan` for ABCD-BIDS, `hcp` for WashU HCPPipelines, `nibabies` for infant-fMRIPrep / NiBabies, and `ukb` for UK Biobank)
```
